File: python/syntonic/nn/winding/syntony.py
"""
Winding Syntony Computer - Compute syntony with winding-aware mode norms.

This module computes syntony S(Ψ) using the winding state structure:

    S(Ψ) = Σᵢ |ψᵢ|² × exp(-|nᵢ|²/φ) / Σᵢ |ψᵢ|²

where |nᵢ|² is the mode norm squared for each feature.
"""

from __future__ import annotations
import torch
import torch.nn as nn
import math

PHI = (1 + math.sqrt(5)) / 2  # Golden ratio ~ 1.618


class WindingSyntonyComputer(nn.Module):
    """
    Computes syntony with winding-aware mode structure.

    The syntony formula weights energy by golden measure:
        S(Ψ) = Σ |ψ_i|² w(nᵢ) / Σ |ψ_i|²

    where:
        w(n) = exp(-|n|²/φ)  # Golden weight

    High syntony (S → 1): Energy concentrated in low-norm modes
    Low syntony (S → 0): Energy scattered across high-norm modes

    Example:
        >>> computer = WindingSyntonyComputer(dim=64)
        >>> x = torch.randn(32, 64)  # Batch of activations
        >>> mode_norms = torch.arange(64).pow(2).float()
        >>> S = computer(x, mode_norms)
        >>> print(f"Syntony: {S:.4f}")
    """

    def __init__(self, dim: int):
        """
        Initialize syntony computer.

        Args:
            dim: Feature dimension
        """
        super().__init__()
        self.dim = dim

    def forward(
        self, x: torch.Tensor, mode_norms: torch.Tensor
    ) -> float:
        """
        Compute syntony over batch.

        Args:
            x: Activations tensor (batch, dim) or (dim,)
            mode_norms: Mode norm squared |n|² for each feature (dim,)

        Returns:
            Scalar syntony S ∈ [0, 1]
        """
        # Ensure x is 2D
        if x.dim() == 1:
            x = x.unsqueeze(0)

        # Energy per feature: |ψᵢ|²
        energy = x.pow(2)  # (batch, dim)

        # Golden weights: w(n) = exp(-|n|²/φ)
        weights = torch.exp(-mode_norms / PHI)  # (dim,)

        # Syntony formula
        weighted_energy = (energy * weights).sum()
        total_energy = energy.sum() + 1e-8  # Avoid division by zero

        syntony = (weighted_energy / total_energy).item()

        # Clamp to [0, 1]
        return max(0.0, min(1.0, syntony))

    def batch_syntony(
        self, x: torch.Tensor, mode_norms: torch.Tensor
    ) -> torch.Tensor:
        """
        Compute syntony for each sample in batch separately.

        Args:
            x: Activations tensor (batch, dim)
            mode_norms: Mode norm squared |n|² for each feature (dim,)

        Returns:
            Syntony values for each sample (batch,)
        """
        # Energy per feature
        energy = x.pow(2)  # (batch, dim)

        # Golden weights
        weights = torch.exp(-mode_norms / PHI)  # (dim,)

        # Syntony per sample
        weighted_energy = (energy * weights).sum(dim=1)  # (batch,)
        total_energy = energy.sum(dim=1) + 1e-8  # (batch,)

        syntony = weighted_energy / total_energy  # (batch,)

        return syntony.clamp(0.0, 1.0)

    def extra_repr(self) -> str:
        return f"dim={self.dim}"


if __name__ == "__main__":
    # Example usage
    computer = WindingSyntonyComputer(dim=64)
    print(f"Syntony computer: {computer}")

    # Create sample data
    x = torch.randn(32, 64)  # Batch of 32
    mode_norms = torch.arange(64).pow(2).float()  # |n|² = 0, 1, 4, 9, 16, ...

    # Compute syntony
    S = computer(x, mode_norms)
    print(f"Batch syntony: {S:.4f}")

    # Per-sample syntony
    S_per_sample = computer.batch_syntony(x, mode_norms)
    print(f"Per-sample syntony (first 5): {S_per_sample[:5]}")
    print(f"Mean: {S_per_sample.mean():.4f}, Std: {S_per_sample.std():.4f}")

    # Test with concentrated energy (low modes)
    x_concentrated = torch.zeros(1, 64)
    x_concentrated[0, :8] = 1.0  # Energy in first 8 modes
    S_high = computer(x_concentrated, mode_norms)
    print(f"\nConcentrated energy syntony: {S_high:.4f}")

    # Test with scattered energy (high modes)
    x_scattered = torch.zeros(1, 64)
    x_scattered[0, -8:] = 1.0  # Energy in last 8 modes
    S_low = computer(x_scattered, mode_norms)
    print(f"Scattered energy syntony: {S_low:.4f}")

File: rust/src/resonant/syntony.rs
//! Syntony computation for winding states.
//!
//! Computes syntony S(Ψ) using winding-aware mode norms:
//!
//!     S(Ψ) = Σᵢ |ψᵢ|² × exp(-|nᵢ|²/φ) / Σᵢ |ψᵢ|²
//!
//! where |nᵢ|² is the mode norm squared for each feature.

use super::number_theory::golden_weight;

/// Compute winding syntony for a 1D tensor.
///
/// S(Ψ) = Σ |ψᵢ|² w(nᵢ) / Σ |ψᵢ|²
///
/// where w(n) = exp(-|n|²/φ) is the golden weight.
///
/// # Arguments
/// * `values` - Tensor values (flattened)
/// * `mode_norms` - Mode norm squared |n|² for each feature
///
/// # Returns
/// Syntony S ∈ [0, 1]
pub fn compute_winding_syntony(values: &[f64], mode_norms: &[f64]) -> f64 {
    if values.is_empty() || mode_norms.is_empty() {
        return 0.0;
    }

    let mut weighted_energy = 0.0;
    let mut total_energy = 0.0;

    for (i, &val) in values.iter().enumerate() {
        let norm = if i < mode_norms.len() {
            mode_norms[i]
        } else {
            // Extrapolate mode norm if fewer norms than values
            mode_norms[mode_norms.len() - 1]
        };
        
        let energy = val * val;
        let weight = golden_weight(norm);
        
        weighted_energy += energy * weight;
        total_energy += energy;
    }

    if total_energy < 1e-10 {
        return 0.0;
    }

    let syntony = weighted_energy / total_energy;
    syntony.clamp(0.0, 1.0)
}

/// Compute batch syntony for a 2D tensor.
///
/// Returns syntony for each sample in the batch.
///
/// # Arguments
/// * `values` - Tensor values (flattened, row-major)
/// * `batch_size` - Number of samples in batch
/// * `dim` - Feature dimension
/// * `mode_norms` - Mode norm squared |n|² for each feature (length = dim)
///
/// # Returns
/// Vector of syntony values, one per sample
pub fn batch_winding_syntony(
    values: &[f64],
    batch_size: usize,
    dim: usize,
    mode_norms: &[f64],
) -> Vec<f64> {
    let mut syntonies = Vec::with_capacity(batch_size);

    // Precompute golden weights
    let weights: Vec<f64> = (0..dim)
        .map(|i| {
            let norm = if i < mode_norms.len() {
                mode_norms[i]
            } else {
                0.0
            };
            golden_weight(norm)
        })
        .collect();

    for b in 0..batch_size {
        let mut weighted_energy = 0.0;
        let mut total_energy = 0.0;

        for d in 0..dim {
            let idx = b * dim + d;
            if idx >= values.len() {
                break;
            }
            
            let val = values[idx];
            let energy = val * val;
            
            weighted_energy += energy * weights[d];
            total_energy += energy;
        }

        let syntony = if total_energy < 1e-10 {
            0.0
        } else {
            (weighted_energy / total_energy).clamp(0.0, 1.0)
        };
        
        syntonies.push(syntony);
    }

    syntonies
}

/// Aggregate syntony using different methods.
///
/// # Arguments
/// * `syntonies` - Per-sample syntony values
/// * `method` - Aggregation method: "mean", "min", "geometric"
///
/// # Returns
/// Aggregated syntony value
pub fn aggregate_syntony(syntonies: &[f64], method: &str) -> f64 {
    if syntonies.is_empty() {
        return 0.5;
    }

    match method {
        "mean" => syntonies.iter().sum::<f64>() / syntonies.len() as f64,
        "min" => syntonies.iter().cloned().fold(f64::INFINITY, f64::min),
        "geometric" => {
            let product: f64 = syntonies.iter()
                .map(|&s| s.max(1e-10))
                .product();
            product.powf(1.0 / syntonies.len() as f64)
        }
        _ => syntonies.iter().sum::<f64>() / syntonies.len() as f64,
    }
}

/// Generate standard mode norms based on index.
///
/// Returns |n|² = i² for each feature index i.
///
/// # Arguments
/// * `dim` - Feature dimension
///
/// # Returns
/// Vector of mode norm squared values
pub fn standard_mode_norms(dim: usize) -> Vec<f64> {
    (0..dim).map(|i| (i * i) as f64).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_compute_winding_syntony() {
        // Concentrated energy in low modes → high syntony
        let values = vec![1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0];
        let mode_norms = standard_mode_norms(8);
        let s = compute_winding_syntony(&values, &mode_norms);
        assert!(s > 0.8, "Expected high syntony for concentrated energy, got {}", s);
    }

    #[test]
    fn test_scattered_energy() {
        // Scattered energy in high modes → low syntony
        let values = vec![0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0];
        let mode_norms = standard_mode_norms(8);
        let s = compute_winding_syntony(&values, &mode_norms);
        // High mode indices (6, 7) have weights exp(-36/φ) and exp(-49/φ) ≈ 0
        assert!(s < 0.1, "Expected low syntony for scattered energy, got {}", s);
    }

    #[test]
    fn test_batch_syntony() {
        // Two samples: one concentrated, one scattered
        let values = vec![
            // Sample 0: concentrated
            1.0, 1.0, 0.0, 0.0,
            // Sample 1: scattered
            0.0, 0.0, 1.0, 1.0,
        ];
        let mode_norms = standard_mode_norms(4);
        let syntonies = batch_winding_syntony(&values, 2, 4, &mode_norms);
        
        assert_eq!(syntonies.len(), 2);
        assert!(syntonies[0] > syntonies[1], "Concentrated should have higher syntony");
    }

    #[test]
    fn test_aggregate_methods() {
        let syntonies = vec![0.6, 0.8, 0.7];
        
        let mean = aggregate_syntony(&syntonies, "mean");
        assert!((mean - 0.7).abs() < 0.001);
        
        let min = aggregate_syntony(&syntonies, "min");
        assert!((min - 0.6).abs() < 0.001);
        
        let geo = aggregate_syntony(&syntonies, "geometric");
        // geometric mean < arithmetic mean
        assert!(geo < mean);
    }
}
